File: maven-projects/java/jmx/arduino-base/src/main/java/br/com/cams7/sisbarc/arduino/status/ArduinoStatus.java
/**
 * 
 */
package br.com.cams7.sisbarc.arduino.status;

import java.util.Properties;

import br.com.cams7.sisbarc.util.AppException;
import br.com.cams7.sisbarc.util.AppUtil;

/**
 * @author cams7
 *
 */
public abstract class ArduinoStatus {

	// Numero maximo da PORTA DIGITAL e 63
	public static final byte DIGITAL_PIN_MAX = 0x3F;
	// Numero maximo da PORTA ANALOGICA e 15
	public static final byte ANALOG_PIN_MAX = 0x0F;

	private static byte[] pinsDigital;
	private static byte[] pinsDigitalPWM;
	private static byte[] pinsAnalog;

	static {
		try {
			Properties pins = AppUtil.getPropertiesFile(ArduinoStatus.class,
					"pins.properties");
			final String SEPARATE = ",";

			pinsDigital = getPins(pins.getProperty("PIN_DIGITAL").trim()
					.split(SEPARATE));
			pinsDigitalPWM = getPins(pins.getProperty("PIN_DIGITAL_PWM").trim()
					.split(SEPARATE));
			pinsAnalog = getPins(pins.getProperty("PIN_ANALOG").trim()
					.split(SEPARATE));
		} catch (AppException e) {
			e.printStackTrace();
		}
	}

	private Transmitter transmitter;
	private Status status;
	private Event event;

	private PinType pinType;

	private byte pin;

	public ArduinoStatus() {
		super();

		setTransmitter(Transmitter.PC);
		setPin((byte) 0x00);
	}

	public ArduinoStatus(Status status, Event event, PinType pinType, byte pin) {
		this();

		setStatus(status);
		setEvent(event);
		setPinType(pinType);
		setPin(pin);
	}

	public void changeCurrentValues(ArduinoStatus arduino) {
		setTransmitter(arduino.getTransmitter());
		setStatus(arduino.getStatus());
		setEvent(arduino.getEvent());

		setPinType(arduino.getPinType());
		setPin(arduino.getPin());
	}

	private static byte[] getPins(final String[] values) {
		byte[] pins = new byte[values.length];
		for (byte i = 0; i < values.length; i++)
			pins[i] = Byte.parseByte(values[i].trim());
		return pins;
	}

	@Override
	public String toString() {
		return this.getClass().getSimpleName() + "[transmitter = "
				+ getTransmitter() + ", status = " + getStatus() + ", event = "
				+ getEvent() + ", pinType = " + getPinType() + ", pin = "
				+ getPin() + "]";
	}

	public Transmitter getTransmitter() {
		return transmitter;
	}

	public void setTransmitter(Transmitter transmitter) {
		this.transmitter = transmitter;
	}

	public Status getStatus() {
		return status;
	}

	public void setStatus(Status status) {
		this.status = status;
	}

	public Event getEvent() {
		return event;
	}

	public void setEvent(Event event) {
		this.event = event;
	}

	public PinType getPinType() {
		return pinType;
	}

	public void setPinType(PinType pinType) {
		this.pinType = pinType;
	}

	public byte getPin() {
		return pin;
	}

	public void setPin(byte pin) {
		this.pin = pin;
	}

	public enum Transmitter {
		ARDUINO, // Mensagem enviada do Arduino
		PC; // Mensagem enviada do PC
	}

	public enum Status {
		SEND, // Mensagem de envio que nao exige uma resposta
		SEND_RESPONSE, // Mensagem de envio que exige uma resposta
		RESPONSE, // Mensagem de resposta
		RESPONSE_RESPONSE;// Mensagem de resposta que exige outra resposta
	}

	public enum Event {
		EXECUTE, WRITE, READ, MESSAGE;
	}

	public enum PinType {
		DIGITAL('d'), // Porta Digital
		ANALOG('a'); // Porta Analogica

		private char type;

		private PinType(char type) {
			this.type = type;
		}

		public char getType() {
			return type;
		}

	}

	public static byte[] getPinsDigital() {
		return pinsDigital;
	}

	public static byte[] getPinsDigitalPWM() {
		return pinsDigitalPWM;
	}

	public static byte[] getPinsAnalog() {
		return pinsAnalog;
	}

}

File: eclipse-workspace/arduino_sisbarc/src/ArduinoStatus.cpp
/*
 * Arduino.cpp
 *
 *  Created on: 21/01/2015
 *      Author: cams7
 */

#include "ArduinoStatus.h"

namespace SISBARC {

const transmitter ArduinoStatus::ARDUINO = 0x00;
const transmitter ArduinoStatus::OTHER_DEVICE = 0x01;

const status ArduinoStatus::SEND = 0x00;
const status ArduinoStatus::SEND_RESPONSE = 0x01;
const status ArduinoStatus::RESPONSE = 0x02;
const status ArduinoStatus::RESPONSE_RESPONSE = 0x03;

const event ArduinoStatus::EXECUTE = 0x00;
const event ArduinoStatus::WRITE = 0x01;
const event ArduinoStatus::READ = 0x02;
const event ArduinoStatus::MESSAGE = 0x03;

const pin_type ArduinoStatus::DIGITAL = 0x00;
const pin_type ArduinoStatus::ANALOG = 0x01;

const uint8_t ArduinoStatus::TRANSMITTER_MAX = OTHER_DEVICE;
const uint8_t ArduinoStatus::STATUS_MAX = RESPONSE_RESPONSE;
const uint8_t ArduinoStatus::EVENT_MAX = MESSAGE;
const uint8_t ArduinoStatus::PIN_TYPE_MAX = ANALOG;

const uint8_t ArduinoStatus::DIGITAL_PIN_MAX = 0x3F; //63
const uint8_t ArduinoStatus::ANALOG_PIN_MAX = 0x0F; //15

const uint8_t ArduinoStatus::PINS_DIGITAL_SIZE = 6;
const uint8_t ArduinoStatus::PINS_DIGITAL[PINS_DIGITAL_SIZE] = { 2, 4, 7, 8, 12,
		13 };

const uint8_t ArduinoStatus::PINS_DIGITAL_PWM_SIZE = 6;
const uint8_t ArduinoStatus::PINS_DIGITAL_PWM[PINS_DIGITAL_PWM_SIZE] = { 3, 5,
		6, 9, 10, 11 };

const uint8_t ArduinoStatus::PINS_ANALOG_SIZE = 6;
const uint8_t ArduinoStatus::PINS_ANALOG[PINS_ANALOG_SIZE] =
		{ 0, 1, 2, 3, 4, 5 };

ArduinoStatus::ArduinoStatus() :
		transmitterValue(ARDUINO), statusValue(SEND), eventValue(EXECUTE), pinType(
				DIGITAL), pin(0x00) {
}

ArduinoStatus::ArduinoStatus(status statusValue, event eventValue,
		pin_type pinType, uint8_t pin) :
		transmitterValue(ARDUINO), statusValue(statusValue), eventValue(
				eventValue), pinType(pinType), pin(pin) {
}

ArduinoStatus::~ArduinoStatus() {
	// TODO Auto-generated destructor stub
}

transmitter ArduinoStatus::getTransmitterValue(void) {
	return transmitterValue;
}

void ArduinoStatus::setTransmitterValue(transmitter transmitterValue) {
	this->transmitterValue = transmitterValue;
}

status ArduinoStatus::getStatusValue(void) {
	return statusValue;
}

void ArduinoStatus::setStatusValue(status statusValue) {
	this->statusValue = statusValue;
}

event ArduinoStatus::getEventValue(void) {
	return eventValue;
}
void ArduinoStatus::setEventValue(event eventValue) {
	this->eventValue = eventValue;
}

pin_type ArduinoStatus::getPinType(void) {
	return pinType;
}

void ArduinoStatus::setPinType(pin_type pinType) {
	this->pinType = pinType;
}

uint8_t ArduinoStatus::getPin(void) {
	return pin;
}

void ArduinoStatus::setPin(uint8_t pin) {
	this->pin = pin;
}

} /* namespace SISBARC */
